pkg/csi/service/vanilla/controller.go

```diff
@@ -1709,14 +1718,240 @@ func (c *controller) createFileVolume(ctx context.Context, req *csi.CreateVolume
+					for _, fsEnabledDatastore := range datastores {
+						for _, ds := range candidateDatastores {
+							if ds.Info.Url == fsEnabledDatastore.Info.Url {
+								fsEnabledCandidateDatastores = append(fsEnabledCandidateDatastores, ds)
+							}
+						}
+					}
+				}
+
+				if scParams.StoragePolicyName != "" {
+					storagePolicyID, err := vcenter.GetStoragePolicyIDByName(ctx, scParams.StoragePolicyName)
+					if err != nil {
+						return nil, csifault.CSIInternalFault, logger.LogNewErrorf(log, "Error occurred while getting Profile Id "+
+							"from Storage Profile Name: %s. Error: %+v", scParams.StoragePolicyName, err)
+					}
+					// Check storage policy compatibility.
+					var sharedDSMoRef []types.ManagedObjectReference
+					for _, ds := range fsEnabledCandidateDatastores {
+						sharedDSMoRef = append(sharedDSMoRef, ds.Reference())
+					}
+					compat, err := vcenter.PbmCheckCompatibility(ctx, sharedDSMoRef, storagePolicyID)
+					if err != nil {
+						return nil, csifault.CSIInternalFault, logger.LogNewErrorf(log, "failed to find datastore compatibility "+
+							"with storage policy ID %q. Error: %+v", storagePolicyID, err)
+					}
+					compatibleDsMoids := make(map[string]struct{})
+					for _, ds := range compat.CompatibleDatastores() {
+						compatibleDsMoids[ds.HubId] = struct{}{}
+					}
+					log.Infof("Datastores compatible with storage policy %q are %+v", scParams.StoragePolicyName,
+						compatibleDsMoids)
+
+					// Filter compatible datastores from shared datastores list.
+					var compatibleDatastores []*cnsvsphere.DatastoreInfo
+					for _, ds := range fsEnabledCandidateDatastores {
+						if _, exists := compatibleDsMoids[ds.Reference().Value]; exists {
+							compatibleDatastores = append(compatibleDatastores, ds)
+						}
+					}
+					if len(compatibleDatastores) == 0 {
+						return nil, csifault.CSIInternalFault, logger.LogNewErrorf(log, "No compatible shared datastores found "+
+							"for storage policy %q", scParams.StoragePolicyName)
+					}
+					fsEnabledCandidateDatastores = compatibleDatastores
+				}
+				log.Infof("fsEnabledCandidateDatastores %v", fsEnabledCandidateDatastores)
+				filterSuspendedDatastores := false
+				if !multivCenterTopologyDeployment {
+					filterSuspendedDatastores = commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx,
+						common.CnsMgrSuspendCreateVolume)
+					if multivCenterCSITopologyEnabled {
+						fsEnabledClusterToDsInfoMap =
+							c.authMgrs[c.managers.CnsConfig.Global.VCenterIP].GetFsEnabledClusterToDsMap(ctx)
+					} else {
+						fsEnabledClusterToDsInfoMap = c.authMgr.GetFsEnabledClusterToDsMap(ctx)
+					}
+				}
+
+				var filteredDatastores []*cnsvsphere.DatastoreInfo
+				if topologyRequirement != nil {
+					filteredDatastores = fsEnabledCandidateDatastores
+				} else {
+					for _, datastores := range fsEnabledClusterToDsInfoMap {
+						filteredDatastores = append(filteredDatastores, datastores...)
```

Comment:
But with this we are overwriting the policy check we did on fsEnabledCandidateDatastores?

Reply:
This code piece is not needed as this is stale. I have removed this if{} else {} block.

Few lines above this code block, we first filter out fsEnabledDatastores & we check if the pbm compatibility passes. So removing this block